shell/platform/common/public/flutter_messenger.h

```diff
@@ -87,6 +88,47 @@ FLUTTER_EXPORT void FlutterDesktopMessengerSetCallback(
     FlutterDesktopMessageCallback callback,
     void* user_data);
 
+// Increments the reference count for the |messenger|.
+//
```

Comment:
We also need to make sure this isn't changing the lifetime in surprising ways. The existing semantics (which need to keep working to avoid breaking people) is that the lifetime of the messenger is (at least as) long as the lifetime of the engine. It looks like the way you've implemented it, if I get a reference, then add a ref, then remove a ref, I've now *destroyed* the messenger, thus shortening its lifespan, which is very unexpected IMO.

Reply:
Added a note the `FlutterDesktopEngineGetMessenger` docstring.

I would consider someone releasing the messenger without retaining it the same problem as someone calling `free` on a returned pointer.  Someone could always go out of their way to fink up the reference count.

Reply:
> I would consider someone releasing the messenger without retaining it the same problem as someone calling `free` on a returned pointer.

But calling `free` is correct if you got the pointer from `malloc`; people can't manage memory correctly if we don't tell them what the semantics of the API are. I think a better example would be whether someone should call `CFRelease` on a Foundation type pointer they received. Imagine trying to use Foundation correctly if the `Create/Copy` rule weren't documented anywhere.

I don't think it's at all obvious that given that there is only one way for someone to get a messenger pointer initially, that one way doesn't give you an owning reference. Either possibility is reasonable absent any documentation.

> Someone could always go out of their way to fink up the reference count.

Of course they can, but we should provide people clear documentation. If people don't know what the the way *is*, it's easy to accidentally go out of it.

Reply:
SG, we are in agreement.  Let's make sure it is clear in the documentation.  I don't think there is anything specified in this thread beyond the `FlutterDesktopEngineGetMessenger` docstring update to change.